### PR TITLE
`General`: Add subgrouping to sidebar for lectures and exercises

### PR DESCRIPTION
**Pull request by eylulnc**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context
This PR introduces consistent **sub‐grouping** for exercises (and lectures). It allows us to display items in sections by weekly buckets when there are more than 5 items.

### Description
- Added a new method to group items by timeframe (past, current and future) and further subdivide them by week if the item count is over 5.
- Displays week the items belongs, if no date provided inside these time frame sit will add a title with  **No Date**.
- No subgrouping happen if the accordion is **No Date**.
- Integrated it into the sidebar accordion for consistent UI/UX.

### Steps for Testing
Prerequisites:: 
Course with at least 6 exercise or lecture in the same section (Current, Past or Future)

1. Log in as an Instructor or Student.
2. Navigate to a course with many exercises (over 5).
3. Verify that exercises/lectures are now subdivided into weekly sections (sub‐headers) if the total is more than 5.

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| course-overview.service.ts | 92.68% | ✅ ❌ |
| sidebar-accordion.component.ts | 95.83% | ✅ ❌ |
| sidebar.ts | not found (modified) | ❌ |
| week-grouping.util.ts | 100% | ✅ |

### Screenshots

Screenshot 2025-04-21 at 22 16 14
Screenshot 2025-04-21 at 20 11 18

## Summary by CodeRabbit

- **New Features**
  - Enhanced lecture cards to include start dates for improved scheduling visibility.
  - Sidebar now organizes items into weekly groups, displaying week ranges for better contextual navigation.

- **Tests**
  - Added comprehensive tests to verify the weekly grouping functionality and correct handling of date-based groupings.

### CODE REVIEW

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Tbh. I'm not entirley sure what's going on there, but it seems the weeks themselves are not orderd by date any more? My guess is, that you're not taking the year into account
image

**Review by asliayk — APPROVED** [review]
tested locally and works as expected 🚀

**Review by FelberMartin — APPROVED** [review]
Nice change, works as described.

One thing I noticed though is that the sorting in Android is the other way round, so when I expand the "Past" section, the oldest exercises are shown first. Is this intended?

**Comment by eylulnc** [comment]
> Nice change, works as described.
> 
> 
> 
> One thing I noticed though is that the sorting in Android is the other way round, so when I expand the "Past" section, the oldest exercises are shown first. Is this intended?

I think its something we should fix in the mobile clients because I based the reference from ios and also android has default order from old to new but web has otherwise so we should make the arrangements on our side to match web. I think this logic makes more sense.

**Review by PaRangger — DISMISSED** [review]
Tested on TS3, works as described. 😄 Code LGTM

**Review by FelberMartin — APPROVED** [review]
Re-approve

**Review by cremertim — DISMISSED** [review]
Code lgmt. Nice changes!

**Review by asliayk — DISMISSED** [review]
reapprove

**Review by muradium — APPROVED** [review]
Code LGTM

**Review by HawKhiem — DISMISSED** [review]
Tested on TS3. Works as expected

image

**Review by krusche — CHANGES_REQUESTED** [review]
The PR currently has merge conflicts

**Review by FelberMartin — DISMISSED** [review]
Re-tested on TS3, works as expected!

**Review by asliayk — APPROVED** [review]
reapprove

**Review by PaRangger — DISMISSED** [review]
Reapprove. Re-tested on TS3.

**Review by HawKhiem — APPROVED** [review]
Reapprove

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts [hunk 1]
```diff
@@ -8,6 +8,17 @@ import { SidebarCardDirective } from '../directive/sidebar-card.directive';
 import { ArtemisTranslatePipe } from 'app/shared/pipes/artemis-translate.pipe';
 import { SearchFilterPipe } from 'app/shared/pipes/search-filter.pipe';
 import { AccordionGroups, ChannelTypeIcons, CollapseState, SidebarCardElement, SidebarItemShowAlways, SidebarTypes } from 'app/shared/types/sidebar';
+import dayjs from 'dayjs/esm';
+import isoWeek from 'dayjs/plugin/isoWeek';
```
Comment (krusche) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts:12]: please move this into dayjs.ts and also use the `.../esm/...` version for the import

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts [hunk 2]
```diff
@@ -93,4 +104,84 @@ export class SidebarAccordionComponent implements OnChanges, OnInit {
...
     }
+
+    private getWeekKey(date: dayjs.Dayjs): string {
+        const weekStart = date.startOf('isoWeek');
+        const weekEnd = date.endOf('isoWeek');
+        const year = weekStart.year();
+        return `${year} - ${weekStart.format('DD MMM YYYY')} - ${weekEnd.format('DD MMM YYYY')}`;
+    }
+
+    getGroupedByWeek(groupKey: string): WeekGroup[] {
```
Comment (krusche) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts:115]: please move this into a util file and make it static or a pure function. This allows to better test the functionality and make sure that all edge cases are covered

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts [hunk 3]
```diff
@@ -93,4 +104,84 @@ export class SidebarAccordionComponent implements OnChanges, OnInit {
...
+            group.push(item);
+            weekGroups.set(weekKey, group);
+        }
+
+        // Sort items within each group by date (newest first)
+        for (const [, items] of weekGroups) {
+            items.sort((a, b) => {
+                const dateA = a.exercise?.dueDate || a.startDateWithTime || a.startDate;
+                const dateB = b.exercise?.dueDate || b.startDateWithTime || b.startDate;
+                if (!dateA && !dateB) return 0;
```
Comment (krusche) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts:155]: please always use curly braces even for one liners

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts [hunk 4]
```diff
@@ -93,4 +104,84 @@ export class SidebarAccordionComponent implements OnChanges, OnInit {
...
+                if (!aFullKey || !bFullKey) return 0;
+
+                const [aYear] = aFullKey.split(' - ');
+                const [bYear] = bFullKey.split(' - ');
+                if (aYear !== bYear) {
+                    return Number(bYear) - Number(aYear);
+                }
+
+                const aDate = dayjs(a.weekRange.split(' - ')[0], 'DD MMM');
+                const bDate = dayjs(b.weekRange.split(' - ')[0], 'DD MMM');
```
Comment (krusche) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts:183]: this code looks quite low-level. You join something with a delimiter and then you split it again. Is there no other more safe mechanism?
If there is really no other way, then at least add some inline documentation, because to me the code is unclear. And make sure to extract `' - '` as a constant to avoid issues and improve maintainability of the code!

**Review by krusche — CHANGES_REQUESTED** [review]
The code is hard to understand and maintain. I suggest you improve it and follow coding best practices. Also make sure you have a full test coverage of the new logic, in particular make sure to take edge cases into account

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts (hunk 4 above)
Comment (eylulnc) [reply src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts]: I have tried to make it better as much as possible 👍🏻

**Review by tobias-lippert — DISMISSED** [review]
code

**Review by PaRangger — DISMISSED** [review]
Re-tested on TS5, everything still works as described. ✅

**Review by asliayk — APPROVED** [review]
retested on ts5

**Review by FelberMartin — DISMISSED** [review]
code

**Review by julian-wls — DISMISSED** [review]
Tested on TS2, works as expected!

**Review by HawKhiem — DISMISSED** [review]
Retested on TS3. Reapprove

**Review by MaximilianAnzinger — DISMISSED** [review]
code

**Review by DenizOzturk95 — DISMISSED** [review]
tested on TS1. worked as described

**Review by HawKhiem — DISMISSED** [review]
Retested on TS1. Approve

image

**Review by krusche — CHANGES_REQUESTED** [review]
please always use curly braces {...} for if, else if and else statements!

**Review by DenizOzturk95 — DISMISSED** [review]
Reapprove manual testing

**Review by HawKhiem — APPROVED** [review]
Reapprove

**Review by tobias-lippert — DISMISSED** [review]
reapprove code

**Review by julian-wls — DISMISSED** [review]
Tested manually, reapprove.

**Review by krusche — DISMISSED** [review]

**Review by krusche — APPROVED** [review]
